File: Labtwo/SWIFT/pr2.swift
import Foundation

// Проверка валидности имени пользователя
func isValidUsername(_ username: String) -> Bool {
    // Проверка длины
    if username.count < 6 || username.count > 30 {
        return false
    }
    
    // Не может начинаться/заканчиваться точкой
    if username.first == "." || username.last == "." {
        return false
    }
    
    // Не может содержать две точки подряд
    if username.contains("..") {
        return false
    }
    
    // Создаем набор допустимых символов
    let validChars = CharacterSet(charactersIn: "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789.*")
    // Проверяем что все символы допустимы
    return username.unicodeScalars.allSatisfy { validChars.contains($0) }
}

// Нормализация email
func normalizeEmail(_ email: String) -> String? {
    // Поиск @
    guard let atPos = email.firstIndex(of: "@") else {
        return nil  // Некорректный email
    }
    
    // Разделение на части
    let username = email[..<atPos]
    let domain = email[atPos...]
    
    var normalized = ""
    var ignoreAfterStar = false
    
    // Обработка каждого символа
    for c in username {
        if c == "*" {
            ignoreAfterStar = true
            continue
        }
        
        // Добавляем символ если не точка и не после звездочки
        if !ignoreAfterStar && c != "." {
            normalized.append(c.lowercased())  // В нижний регистр
        }
    }
    
    // Возвращаем нормализованный email
    return normalized + domain.lowercased()
}

// Основная программа
print("Enter emails (separated by commas):")
guard let input = readLine() else {
    exit(0)
}

// Множество для хранения уникальных email
var uniqueEmails = Set<String>()

// Обработка каждого email
for email in input.components(separatedBy: ",") {
    let trimmedEmail = email.trimmingCharacters(in: .whitespacesAndNewlines)
    if trimmedEmail.isEmpty {
        continue
    }
    
    // Поиск @
    if let atPos = trimmedEmail.firstIndex(of: "@") {
        let username = String(trimmedEmail[..<atPos])
        // Проверка валидности и нормализация
        if isValidUsername(username) {
            if let normalized = normalizeEmail(trimmedEmail) {
                uniqueEmails.insert(normalized)  // Добавление в множество
            }
        }
    }
}

// Вывод результата
print("Number of unique emails: \(uniqueEmails.count)")

File: Labtwo/JAVA/pr2.java
import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;

public class EmailValidator {
    /**
     * Проверка валидности имени пользователя
     * @param username - имя пользователя для проверки
     * @return true если имя валидно, иначе false
     */
    public static boolean isValidUsername(String username) {
        // Длина имени должна быть от 6 до 30 символов
        if (username.length() < 6 || username.length() > 30) {
            return false;
        }

        // Не может начинаться или заканчиваться точкой
        if (username.startsWith(".") || username.endsWith(".")) {
            return false;
        }

        // Не может содержать две точки подряд
        if (username.contains("..")) {
            return false;
        }

        // Может содержать только буквы, цифры, точки и звездочки
        for (char c : username.toCharArray()) {
            if (!(Character.isLetterOrDigit(c) || c == '.' || c == '*')) {
                return false;
            }
        }

        return true;
    }

    /**
     * Нормализация email адреса
     * @param email - email для нормализации
     * @return нормализованный email или null если email невалиден
     */
    public static String normalizeEmail(String email) {
        int atIndex = email.indexOf('@');
        if (atIndex == -1) {
            return null; // Некорректный email (нет @)
        }

        String username = email.substring(0, atIndex);
        String domain = email.substring(atIndex).toLowerCase(); // Домен в нижний регистр

        // Проверяем валидность имени пользователя
        if (!isValidUsername(username)) {
            return null;
        }

        StringBuilder normalized = new StringBuilder();
        boolean ignoreAfterStar = false;

        // Обрабатываем имя пользователя
        for (char c : username.toCharArray()) {
            if (c == '*') {
                ignoreAfterStar = true;
                continue;
            }

            if (!ignoreAfterStar && c != '.') {
                normalized.append(Character.toLowerCase(c));
            }
        }

        // Возвращаем нормализованный email
        return normalized.toString() + domain;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Set<String> uniqueEmails = new HashSet<>();

        System.out.println("Enter emails separated by commas:");
        String input = scanner.nextLine();

        // Разделяем ввод по запятым
        String[] emails = input.split(",");

        for (String email : emails) {
            String trimmedEmail = email.trim();
            if (!trimmedEmail.isEmpty()) {
                String normalized = normalizeEmail(trimmedEmail);
                if (normalized != null) {
                    uniqueEmails.add(normalized);
                }
            }
        }

        System.out.println("Number of unique valid emails: " + uniqueEmails.size());
        scanner.close();
    }
}